test/gtest/uct/ib/test_devx.cc

```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
+     */
+    bool check_xgvmi_effective() const {
+        if (check_xgvmi()) {
+            uct_ib_mlx5_devx_mem_t *memh = create_memh(1);
+            ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+            /* XGVMI capability might be dropped by this point! */
+            destroy_memh(memh);
+            return check_xgvmi();
+        }
+
+        return false;
+    }
+
+    uct_ib_mlx5_devx_mem_t *create_memh(size_t length) const {
+        void *addr                     = ucs_malloc(length, "test");
+        uct_md_mem_reg_params_t params = {};
+        uct_mem_h memh;
+
+        EXPECT_NE(nullptr, addr);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_reg(m_e->md(), addr, length, &params,
+                                               &memh));
+        return ucs_derived_of(memh, uct_ib_mlx5_devx_mem_t);
+    }
+
+    void destroy_memh(uct_ib_mlx5_devx_mem_t *memh) const {
+        void *addr                       = memh->address;
+        uct_md_mem_dereg_params_t params = {};
+
+        params.field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH;
+        params.memh       = (uct_mem_h)memh;
+
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_dereg(m_e->md(), &params));
+
+        if (nullptr != addr) {
+            ucs_free(addr);
+        }
+    }
+
+    void check_umr_init(bool init_export, bool init_import) const {
+        EXPECT_EQ(init_export, nullptr != md()->umr.qp);
+        EXPECT_EQ(init_export, nullptr != md()->umr.cq);
+        if (!init_export) {
+            EXPECT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
+        }
+
+        EXPECT_EQ(init_import, nullptr != md()->umr.mkey_hash);
+    }
+
+    static void check_memh_export(uct_ib_mlx5_devx_mem_t *memh, bool is_umr) {
+        EXPECT_EQ(is_umr, UCT_IB_MLX5_MKEY_TAG_UMR ==
+                          (memh->exported_lkey & 0xff));
+        EXPECT_EQ(is_umr, nullptr != memh->exported_umr_mkey);
+        EXPECT_EQ(is_umr, nullptr == memh->cross_mr);
+    }
+};
+
+UCS_TEST_P(test_devx_umr_mkey, lazy_init)
+{
+    check_umr_init(false, false);
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_no_xgvmi, !check_xgvmi())
+{
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+
+    /* Drop XGVMI capability */
+    md()->flags &= ~UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;
+
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+
+    /* UMR objects remains uninitialized, exported key is not UMR one */
+    check_umr_init(false, false);
+    check_memh_export(memh, false);
+    destroy_memh(memh);
+}
+
+UCS_TEST_SKIP_COND_P(test_devx_umr_mkey, export_mkey_xgvmi, !check_xgvmi())
+{
+    uct_ib_mlx5_devx_mem_t *memh = create_memh(1024);
+    ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+    /* XGVMI capability might be dropped by this point! */
+
+    /* UMR export transport objects must be initialized */
+    check_umr_init(true, false);
+    /* UMR mkey pool must be empty */
+    ASSERT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
```

Comment:
why is it important the mkey pool is empty?

Reply:
would ASAN catch it if we had a memory leak?

Reply:
Well, if we had a true memory leak, it must be catch by LSAN.
Here however I don't check for a memory leak. All the objects stored in the mkey pool or alias hash map will be destroyed at the end of the process, so no memory leak will be detected in this particular case. I just validate the invariant: pool must not grow up on export.
Again, I don't consider this check as super important, so we can remove it if you dislike it